s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -0,0 +1,245 @@
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.AsciiString;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import javax.annotation.Nullable;
+
+/** Factory for performing negotiation of a secure channel using the S2A. */
+@ThreadSafe
+public final class S2AProtocolNegotiatorFactory {
+  @VisibleForTesting static final int DEFAULT_PORT = 443;
+  private static final AsciiString SCHEME = AsciiString.of("https");
+
+  /**
+   * Creates a {@code S2AProtocolNegotiatorFactory} configured for a client to establish secure
+   * connections using the S2A.
+   *
+   * @param localIdentity the identity of the client; if none is provided, the S2A will use the
+   *     client's default identity.
+   * @param s2aChannelPool a pool of shared channels that can be used to connect to the S2A.
+   * @return a factory for creating a client-side protocol negotiator.
+   */
+  public static InternalProtocolNegotiator.ClientFactory createClientFactory(
+      Optional<S2AIdentity> localIdentity, ObjectPool<Channel> s2aChannelPool) {
+    checkNotNull(s2aChannelPool, "S2A channel pool should not be null.");
+    checkNotNull(localIdentity, "Local identity should not be null on the client side.");
+    S2AChannelPool channelPool = S2AGrpcChannelPool.create(s2aChannelPool);
+    return new S2AClientProtocolNegotiatorFactory(localIdentity, channelPool);
+  }
+
+  static final class S2AClientProtocolNegotiatorFactory
+      implements InternalProtocolNegotiator.ClientFactory {
+    private final Optional<S2AIdentity> localIdentity;
+    private final S2AChannelPool channelPool;
+
+    S2AClientProtocolNegotiatorFactory(
+        Optional<S2AIdentity> localIdentity, S2AChannelPool channelPool) {
+      this.localIdentity = localIdentity;
+      this.channelPool = channelPool;
+    }
+
+    @Override
+    public ProtocolNegotiator newNegotiator() {
+      return S2AProtocolNegotiator.createForClient(channelPool, localIdentity);
+    }
+
+    @Override
+    public int getDefaultPort() {
+      return DEFAULT_PORT;
+    }
+  }
+
+  /** Negotiates the TLS handshake using S2A. */
+  @VisibleForTesting
+  static final class S2AProtocolNegotiator implements ProtocolNegotiator {
+
+    private final S2AChannelPool channelPool;
+    private final Optional<S2AIdentity> localIdentity;
+    private final ListeningExecutorService service =
+        MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(1));
+
+    static S2AProtocolNegotiator createForClient(
+        S2AChannelPool channelPool, Optional<S2AIdentity> localIdentity) {
+      checkNotNull(channelPool, "Channel pool should not be null.");
+      checkNotNull(localIdentity, "Local identity should not be null on the client side.");
```

Comment:
Correct, however I think the reason we Optional<S2AIdentity> is because the localIdentity can be empty, in this case S2A will choose a default localIdentity, see https://github.com/grpc/grpc-java/pull/11113/files#diff-7502f77f0608fe8242cff3b41e9c4176bc362d2ca191a0b4a9121b7e4f9496f6R104-R106, for example.

cc: @matthewstevenson88

Reply:
My point was that you should either reject Optional.empty values like you are null values, or if it is okay to have empty values, then just convert nulls to Optional.empty instead of throwing an IllegalArgumentException.  There are a number of places like this where you do a check for null on an Optional.

Reply:
Thanks for explaining, this makes sense. It is ok to have Optional.empty, so I have pushed down the creation of the actual Optional.empty from a null to S2AProtocolNegotiator. Changes in: https://github.com/grpc/grpc-java/pull/11113/commits/b021a2128ec2dc92fa403c190d4aeaa60c259a78